File: 496_Next_Greater_Element_I.cs
public class Solution {
    public int[] NextGreaterElement(int[] nums1, int[] nums2) {
        var dic = new Dictionary<int, int>();
        var s = new Stack<int>();
        foreach (var num in nums2) {
            while (s.Count() > 0 && s.Peek() < num) {
                dic[s.Pop()] = num;
            }
            s.Push(num);
        }
        for (var i = 0; i < nums1.Count(); ++i) {
            nums1[i] = dic.ContainsKey(nums1[i]) ? dic[nums1[i]] : -1;
        }
        return nums1;
    }
}

File: 496_Next_Greater_Element_I.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        s, d = [], {}
        for i in range(len(nums2))[::-1]:
            while len(s) and nums2[s[-1]] < nums2[i]:
                s.pop()
            d[nums2[i]] = nums2[s[-1]] if len(s) else -1
            s.append(i)
        return [d[x] for x in nums1]
